runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,481 @@
+}
+
+// Obtains the minimum epoch of all threads.
+static uint64_t safe_epoch(void);
+static uint64_t safe_epoch(void) {
+  uint64_t min = (uint64_t) -1;
+  // Check all remaining threads for whether they have updated to a more recent epoch.
+  for (struct tls_node *node = get_tls_list(); node != NULL; node = node->next) {
+    if (get_tasks(node) > 0) {
+      uint64_t epoch = get_epoch(node);
+      min = min ? (min < epoch) : epoch;       
+    }
+  }
+  
+  return min;
+}
+
+// Requires spinlock.
+static inline void handle_deferred_data(struct tls_node *node);
+static inline void handle_deferred_data(struct tls_node *node) {
+  // Acquire all data that can be deleted.
+  uint64_t epoch = safe_epoch();
+  struct defer_node *list = pop_bulk_defer_list(node, epoch);
+
+  // Handle deletion of said data...
+  for (struct defer_node *dnode = list; dnode != NULL;) {
+    struct defer_node *tmp = dnode;
+    dnode = dnode->next;
+    delete_data(tmp);
+    push_recycle_list(node, tmp);
+  }
+}
+
+void chpl_qsbr_init(void) {
+    CHPL_TLS_INIT(chpl_qsbr_tls);
+}
+
+void chpl_qsbr_onTaskCreation(void) {
+  struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  if (tls == NULL) {
+    init_tls();
+    tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  }
+
+  // Notify that we are now in-use.
+  add_task(tls);
+}
+
+
+void chpl_qsbr_onTaskDestruction(void) {
+  struct tls_node *tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  if (tls == NULL) {
+    init_tls();
+    tls = CHPL_TLS_GET(chpl_qsbr_tls);
+  }
+
+  // Notify that we have one less task
+  uint64_t nTasks = remove_task(tls) - 1;
+
+  // If we are the last task, handle any extra deferred data...
+  if (nTasks == 0) {
+    acquire_spinlock(tls);
+
+    // Empty as best we can
+    handle_deferred_data(tls);
+    if (tls->deferList == NULL) {
+      release_spinlock(tls);
+      return;
+    }
+
+    // At this point if we have any deferred data, we must pass this to some
+    // other thread, as we could be parked indefinitely. Find first active
+    // thread to dump the rest of our work on...
+    struct tls_node *node = get_tls_list();
+    for (; node != NULL; node = node->next) {
+      if (get_tasks(node) > 0) {
+        // Double check...
+        acquire_spinlock(node);
+        if (get_tasks(node) > 0) {
+          break;
+        }
+        release_spinlock(node);  
+      }
+    }
+
+    // If node is NULL, then all threads are finished so we should be able to safely
+    // handle our deferred data...
+    if (node == NULL) {
+      handle_deferred_data(tls);
+    } else {
+      // If the node is not NULL, we found our thread to dump on and we already hold its lock.
+      struct defer_node *list = pop_all_defer_list(tls);
+      
+      // If their list is empty its easy
+      if (node->deferList == NULL) {
+        node->deferList = list;
+      } else {
+        // TODO: Optimize! If tls->deferList is size N and node->deferList is size M,
+        // then is O((M+N) * N) = O(M*N^2).
```

Comment:
Isn't the optimized version just merging two sorted lists? Seems like something easy enough to write that you should do it before we merge this PR. Anyway I'd appreciate a comment like "Merge the lists, but keep them sorted by epoch".

Reply:
Oh yeah, I can create a new sorted list rather than modifying in-place, that would be easy, lol.